### ISSUE TITLE
Methods don't get generated in the Live part

### ISSUE THREAD

**Issue by guydols** [NimonSour/interthread#7]
I'm running into the error `method not found in ActorLive` in some project. Recreated the issue in the form of the example of the library. I looked at the generated code and, indeed, for some reason the Live part does not contain the generated methods.

The example works just fine, but I can't seem to figure out why I won't with the code below.

```rust
pub struct Actor {
    database: String,
}

#[interthread::actor]
impl Actor {
    pub fn new(database: String) -> Self {
        Self { database }
    }
    pub fn bump() {}
    pub fn get_user(self, username: String) -> String {
        "read_user".into()
    }
}
fn main() {

    let actor = 
    ActorLive::new("test".to_owned());

    let mut actor_a = actor.clone();
    let mut actor_b = actor.clone();

    actor_a.bump();

    let handle_a = std::thread::spawn( move || { 
    actor_a.get_user("test");
    });

    let handle_b = std::thread::spawn( move || {
    actor_b.get_user("test")
    });

    let _  = handle_a.join();
    let hb = handle_b.join().unwrap();
}
```

Using crates:
interthread = "1.2.4"
oneshot     = "0.1.6" 

Rust version:
cargo 1.77.2 (e52e36006 2024-03-26)

Thanks for the effort!

**Comment by NimonSour** [comment]
Hi, Guy Dols !
The behaviour you're observing is indeed intentional ( doesn't mean it's nice one).

The macro is designed to target methods that utilize as receiver 
either `&self` or `&mut self` ( I will improve the documentation, just to make sure that the reader can't avoid this statement).

Lets look at what you're pointing at:

`pub fn bump()` - this is a static method has no receiver at all, there is no way to read 
or mutate the state, the macro will ignore it ( this will change in future versions ).

`pub fn get_user(self, username: String) -> String` - 
this method has a `self` receiver which consumes the instance, this is against the model itself, no instance - no model.

In the first case, it was solely my mistake as I was unable to provide 
a straightforward justification at the time for:

1. Why we need Actor's associated (static) methods in the model? 

2. What associated methods should be included in the model?

See the option  [`assoc`](https://docs.rs/interthread/1.2.4/interthread/attr.actor.html#assoc) it makes the macro to take in consideration public static methods. But 
it will not work for in this case as there isn't any input or output ( so it will be ignored anyway ) .. as I've said my mess.

The plan is for the next release ( hopefully soon as I'm a bit bussy at the moment ) to implement some 
changes which will remove the [`assoc`](https://docs.rs/interthread/1.2.4/interthread/attr.actor.html#assoc) option in favor of default inclusion of static methods
based on following arguments : 
1. We need associated methods in the model because it saves from exporting ( declaring public ) Actor type (and then importing it) just for the sake of using some associated methods on it. 
2. All non-private (pub , pub(crate) and so on.. ). 

This behaviour will change in future release.


Next, second  method `get_user` consumes `self`, the model simply can't be reliably constructed for methods that consume the actor instance, therefore the macro will ignore these methods
letting the user deal with them if any.

I recognize that the current behavior may appear ignorant at first glance.
However, raising errors for unprocessable code, 
will create impediments for users who're customizing the model to meet their 
specific requirements by adding necessary functionalities. 


Even if the method signature will be of the form `pub fn get_user(self, username: String) -> Self`
we can detect it and assume that it's a possible swap, note that in this case we will not return 
anything to ActorLive instance. So we can perform just a swap.. but a simple swap of the Actor doesn't require the method to consume the instance like in this example: 


```rust 

struct Actor(String);

#[interthread::actor]
impl Actor {

    pub fn new( name: &'static str ) -> Self {
        Self(name.to_string())
    }

    pub fn get_name(&self) -> String {
        self.0.clone()
    }

    pub fn swap(&mut self, other: Self ) -> Self {
        std::mem::replace( self, other)
    }
} 


fn main() {

    let mut act_one = ActorLive::new("One");
    // check name
    println!("First  - {}", act_one.get_name());

    let act_two = Actor::new("Two");
    // swap the actors
    let _ = act_one.swap(act_two);

    // check name
    println!("Second - {}", act_one.get_name());

    // we need to wait 
    std::thread::sleep(std::time::Duration::from_secs(1));

}

```
Going back to original question (design),
if you where to trick the macro by changing the method signature receiver to an acceptable one ( `&self` or `&mut self` ),
and then print it on the file via [`edit`](https://docs.rs/interthread/1.2.4/interthread/attr.actor.html#edit) and [`file`](https://docs.rs/interthread/1.2.4/interthread/attr.actor.html#file)  and then changing the receiver back to 
`self`, Rust will not compile the code as it's capable to detect the trouble, we need to do some additional changes in `ActorScript::direct` :

```rust

pub struct Actor {
    database: String,
}

#[interthread::actor(file="path/to/you_file.rs",
// edit(script,live(imp(get_user))))  ]
edit(script))]

impl Actor {
    pub fn new(database: String) -> Self {
        Self { database }
    }
    pub fn method_1(&self) {}
    pub fn method_2(&self) {}
    pub fn get_user(self, username: String) -> String {
        "read_user".into()
    }
}

//++++++++++++++++++[ Interthread  Write to File ]+++++++++++++++++//
// Object Name   : Actor  
// Initiated By  : #[interthread::actor(file="path/to/you_file.rs",edit(file(script),live(imp(file(get_user)))))]  


pub enum ActorScript {
    Method1 {},
    Method2 {},
    GetUser { input: (String), inter_send: oneshot::Sender<String> },
}
impl ActorScript {
    pub fn direct(self, actor: &mut Actor) {
        match self {
            ActorScript::Method1 {} => {
                actor.method_1();
            }
            ActorScript::Method2 {} => {
                actor.method_2();
            }
            ActorScript::GetUser { input: (username), inter_send } => {
                // we may create and swap a new actor for every call to "get_user"
                //let old_actor = std::mem::replace( actor, Actor::new( "New".to_string()) );
                
                inter_send
                    //.send(old_actor.get_user(username))
                    .send(actor.get_user(username))
                    .unwrap_or_else(|_error| {
                        core::panic!(
                            "'ActorScript::GetUser.direct'. Sending on a closed channel."
                        )
                    });
            }
        }
    }
    pub fn play(receiver: std::sync::mpsc::Receiver<ActorScript>, mut actor: Actor) {
        while let std::result::Result::Ok(msg) = receiver.recv() {
            msg.direct(&mut actor);
        }
        eprintln!("Actor the end ...");
    }
}
impl std::fmt::Debug for ActorScript {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ActorScript::Method1 { .. } => write!(f, "ActorScript::Method1"),
            ActorScript::Method2 { .. } => write!(f, "ActorScript::Method2"),
            ActorScript::GetUser { .. } => write!(f, "ActorScript::GetUser"),
        }
    }
}
impl ActorLive {
    pub fn get_user(&self, username: String) -> String {
        let (inter_send, inter_recv) = oneshot::channel();
        let msg = ActorScript::GetUser {
            input: (username),
            inter_send,
        };
        let _ = self
            .sender
            .send(msg)
            .expect("'ActorLive::method.send'. Channel is closed!");
        inter_recv
            .recv()
            .unwrap_or_else(|_error| {
                core::panic!("'Actor::get_user' from inter_recv. Channel is closed!")
            })
    }
}

// *///.............[ Interthread  End of Write  ].................//


fn main() {

    let mut act = ActorLive::new("MyInstance".to_string());

    let _ = act.get_user("Bla".to_string());
}

```

It has to be done manually case dependent.

So to recap, I will change static methods soon but 'self' receivers 
I believe it's preferable to leave on user's experience as there isn't a 
generalized solution.

What are your thoughts on this matter?

Thank you for your involvement!

**Comment by guydols** [comment]
Hey Simon, 

Thank you for the time and effort in explaining the design choices.
It's much clearer now with the examples you have given.

I was able to rectify it in the project I was working on when I stumbled onto the problem. In hindsight, I feel kind of silly not seeing the pattern when compared to the library example. But now that I have a better understanding of the design, it makes perfect sense. 

BTW, great library, I often dreamed of this kind of solution to the problem of having to manually define channels and interactions over and over again.

Great work!
